Go: sync ExternalFlow.qll

owen-mc opened the pull request:
Update go's copy of `ExternalFlow.qll` to bring it in line with java and csharp's.

- Add support for neutral models.
- Change the way that the receiver is referred to in models-as-data models. Previously it was `Argument[-1]`. Now it is `Argument[receiver]`. (Other languages use different terminology, such as `Argument[this]`.)

smowton commented on go/ql/lib/semmle/go/dataflow/ExternalFlow.qll:
```diff
@@ -96,28 +95,18 @@ private predicate packageLink(string shortpkg, string longpkg) {
  * Holds if MaD framework coverage of `package` is `n` api endpoints of the
- * kind `(kind, part)`, and `pkgs` is the number of subpackages of `package`
- * which have MaD framework coverage (including `package` itself).
+ * kind `(kind, part)`.
```
Why does this stop documenting `pkgs`, which still exists?

owen-mc replied:
To match the java and C# versions of this file, so they're easier to keep in sync or make into a shared library. You're right that it's better though. Maybe I'll keep it and make a tiny PR to add it to the others.

owen-mc replied:
Done here: https://github.com/github/codeql/pull/16676

smowton commented on go/ql/lib/semmle/go/dataflow/ExternalFlow.qll:
```diff
@@ -434,6 +434,10 @@ private class SummarizedCallableAdapter extends SummarizedCallable {
+    ) and
+    not exists(Provenance provenance |
+      neutralElement(this, "summary", provenance) and
+      provenance.isManual()
```
isGenerated

owen-mc replied:
Actually it is correct. The idea, which could do with a comment explaining it, is that generated summary models are overridden by manual neutral models. This allows us to fixed generated models in a way which will persist when they are regenerated.

owen-mc replied:
Let me know if you think there should be a comment briefly explaining this (and feel free to give a suggestion).

smowton commented:
Looks plausible, except as noted

owen-mc commented:
I've addressed the review comments and rebased to check that this is still in a mergeable state, since it was quite old. There has been one change to the java copy of `ExternalFlow.qll` since I made the PR but it isn't relevant to go so I'll leave it for now.
